File: AIQuotaBar/Services/Clash/ClashConfigurationDiscovery.swift
import Foundation

struct ClashConfigurationDiscovery {
    private let fileManager: FileManager
    private let homeDirectory: URL

    init(
        fileManager: FileManager = .default,
        homeDirectory: URL = FileManager.default.homeDirectoryForCurrentUser
    ) {
        self.fileManager = fileManager
        self.homeDirectory = homeDirectory
    }

    func discover() throws -> ClashControllerConfiguration {
        var foundConfiguration = false

        for candidate in candidateConfigurations {
            guard fileManager.fileExists(atPath: candidate.url.path) else { continue }
            foundConfiguration = true

            guard let data = fileManager.contents(atPath: candidate.url.path),
                  let contents = String(data: data, encoding: .utf8) else {
                continue
            }

            let values = ClashTopLevelYAML.parse(contents)
            guard let rawController = values["external-controller"]?
                .trimmingCharacters(in: .whitespacesAndNewlines),
                !rawController.isEmpty else {
                continue
            }

            let baseURL = try Self.controllerURL(from: rawController)
            return ClashControllerConfiguration(
                baseURL: baseURL,
                secret: values["secret"] ?? "",
                clientName: candidate.clientName,
                configURL: candidate.url)
        }

        if foundConfiguration {
            throw ClashIntegrationError.externalControllerDisabled
        }
        throw ClashIntegrationError.configurationNotFound
    }

    private var candidateConfigurations: [(url: URL, clientName: String)] {
        [
            (
                homeDirectory.appending(
                    path: "Library/Application Support/io.github.clash-verge-rev.clash-verge-rev/clash-verge.yaml"),
                "Clash Verge Rev"
            ),
            (
                homeDirectory.appending(path: ".config/mihomo/config.yaml"),
                "Mihomo"
            ),
            (
                homeDirectory.appending(path: ".config/clash/config.yaml"),
                "Clash"
            ),
        ]
    }

    static func controllerURL(from rawAddress: String) throws -> URL {
        var address = rawAddress.trimmingCharacters(in: .whitespacesAndNewlines)
        if address.hasPrefix(":") {
            address = "127.0.0.1\(address)"
        } else if address.hasPrefix("*:") {
            address = "127.0.0.1:\(address.dropFirst(2))"
        } else if address.hasPrefix("0.0.0.0:") {
            address = "127.0.0.1:\(address.dropFirst("0.0.0.0:".count))"
        } else if address.hasPrefix("[::]:") {
            address = "127.0.0.1:\(address.dropFirst("[::]:".count))"
        }

        if !address.contains("://") {
            address = "http://\(address)"
        }

        guard let components = URLComponents(string: address),
              let host = components.host?.lowercased(),
              let port = components.port,
              port > 0,
              let url = components.url else {
            throw ClashIntegrationError.invalidControllerAddress(rawAddress)
        }

        let localHosts = Set(["127.0.0.1", "localhost", "::1"])
        guard localHosts.contains(host) else {
            throw ClashIntegrationError.unsafeControllerHost(host)
        }

        return url
    }
}

enum ClashTopLevelYAML {
    static func parse(_ contents: String) -> [String: String] {
        var result: [String: String] = [:]

        for rawLine in contents.components(separatedBy: .newlines) {
            guard let first = rawLine.first,
                  !first.isWhitespace,
                  first != "#",
                  let separator = rawLine.firstIndex(of: ":") else {
                continue
            }

            let key = rawLine[..<separator].trimmingCharacters(in: .whitespaces)
            guard !key.isEmpty else { continue }

            let valueStart = rawLine.index(after: separator)
            let rawValue = String(rawLine[valueStart...])
            result[key] = decodeScalar(rawValue)
        }

        return result
    }

    private static func decodeScalar(_ rawValue: String) -> String {
        let trimmed = rawValue.trimmingCharacters(in: .whitespaces)
        guard !trimmed.isEmpty else { return "" }

        if trimmed.hasPrefix("'") {
            guard let closingIndex = closingSingleQuote(in: trimmed) else {
                return String(trimmed.dropFirst())
            }
            let value = trimmed[trimmed.index(after: trimmed.startIndex)..<closingIndex]
            return value.replacingOccurrences(of: "''", with: "'")
        }

        if trimmed.hasPrefix("\"") {
            if let data = trimmed.data(using: .utf8),
               let decoded = try? JSONDecoder().decode(String.self, from: data) {
                return decoded
            }
            if let closingIndex = closingDoubleQuote(in: trimmed) {
                return String(trimmed[trimmed.index(after: trimmed.startIndex)..<closingIndex])
            }
        }

        if let commentRange = trimmed.range(of: " #") {
            return String(trimmed[..<commentRange.lowerBound])
                .trimmingCharacters(in: .whitespaces)
        }
        return trimmed
    }

    private static func closingSingleQuote(in value: String) -> String.Index? {
        var index = value.index(after: value.startIndex)
        while index < value.endIndex {
            guard value[index] == "'" else {
                index = value.index(after: index)
                continue
            }

            let next = value.index(after: index)
            if next < value.endIndex, value[next] == "'" {
                index = value.index(after: next)
                continue
            }
            return index
        }
        return nil
    }

    private static func closingDoubleQuote(in value: String) -> String.Index? {
        var index = value.index(after: value.startIndex)
        var isEscaped = false

        while index < value.endIndex {
            let character = value[index]
            if character == "\"", !isEscaped {
                return index
            }
            if character == "\\" {
                isEscaped.toggle()
            } else {
                isEscaped = false
            }
            index = value.index(after: index)
        }
        return nil
    }
}

File: windows/src/AIQuotaBar.Providers/Clash/ClashConfigurationDiscovery.cs
// Swift 来源：AIQuotaBar/Services/Clash/ClashConfigurationDiscovery.swift — struct ClashConfigurationDiscovery（v1.28.1）
// 对应测试：AIQuotaBar/Tests/Clash/ClashConfigurationDiscoveryTests.swift
// fixtures：windows/contracts/fixtures/clash/clash-verge-config.yaml
//
// Windows 路径语义（相对 Swift 的差异，发现逻辑结构不变）：
// - Clash Verge Rev 数据目录：macOS 为 ~/Library/Application Support/io.github.clash-verge-rev.clash-verge-rev，
//   Windows 为 %APPDATA%\io.github.clash-verge-rev.clash-verge-rev（另保留 %APPDATA%\clash-verge-rev 变体）；
// - Mihomo / Clash 仍走 %USERPROFILE%\.config\{mihomo,clash}\config.yaml。

#nullable enable

using System;
using System.Collections.Generic;
using System.IO;
using System.Threading;
using System.Threading.Tasks;

namespace AIQuotaBar.Providers.Clash;

/// <summary>按候选顺序发现本机 Clash 配置并解析出控制器连接参数。</summary>
public sealed class ClashConfigurationDiscovery
{
    private readonly IClashEnvironment _environment;

    public ClashConfigurationDiscovery(IClashEnvironment? environment = null)
    {
        _environment = environment ?? ClashEnvironment.Instance;
    }

    public async Task<ClashControllerConfiguration> DiscoverAsync(CancellationToken cancellationToken = default)
    {
        var foundConfiguration = false;

        foreach (var candidate in CandidateConfigurations())
        {
            if (!File.Exists(candidate.Path))
            {
                continue;
            }

            foundConfiguration = true;

            string contents;
            try
            {
                contents = await File.ReadAllTextAsync(candidate.Path, cancellationToken).ConfigureAwait(false);
            }
            catch (Exception exception) when (exception is IOException or UnauthorizedAccessException)
            {
                continue;
            }

            var values = ClashTopLevelYaml.Parse(contents);
            if (values.TryGetValue("external-controller", out var rawController) &&
                rawController.Trim().Length > 0)
            {
                var baseUrl = ControllerUrl(rawController);
                return new ClashControllerConfiguration(
                    BaseUrl: baseUrl,
                    Secret: values.TryGetValue("secret", out var secret) ? secret : string.Empty,
                    ClientName: candidate.ClientName,
                    ConfigUrl: new Uri(candidate.Path));
            }
        }

        if (foundConfiguration)
        {
            throw new ClashIntegrationException(new ClashIntegrationError.ExternalControllerDisabled());
        }

        throw new ClashIntegrationException(new ClashIntegrationError.ConfigurationNotFound());
    }

    private IReadOnlyList<CandidateConfiguration> CandidateConfigurations()
    {
        return new[]
        {
            new CandidateConfiguration(
                Path.Combine(
                    _environment.AppData,
                    "io.github.clash-verge-rev.clash-verge-rev",
                    "clash-verge.yaml"),
                "Clash Verge Rev"),
            new CandidateConfiguration(
                Path.Combine(_environment.AppData, "clash-verge-rev", "clash-verge.yaml"),
                "Clash Verge Rev"),
            new CandidateConfiguration(
                Path.Combine(_environment.Home, ".config", "mihomo", "config.yaml"),
                "Mihomo"),
            new CandidateConfiguration(
                Path.Combine(_environment.Home, ".config", "clash", "config.yaml"),
                "Clash"),
        };
    }

    /// <summary>
    /// 归一化 external-controller 地址并强制**仅回环**安全校验：
    /// ":9097" / "*:9097" / "0.0.0.0:9097" / "[::]:9097" 统一归一为 127.0.0.1，
    /// 无 scheme 时补 http://；主机必须是 127.0.0.1 / localhost / ::1 且端口显式存在，
    /// 否则抛 UnsafeControllerHost / InvalidControllerAddress。
    /// 产出为 Swift URL.absoluteString 字符串形态（scheme://host:port，无尾斜杠；
    /// .NET Uri 对 authority-only URI 会规范出尾斜杠，故不走 ToString/AbsoluteUri）。
    /// </summary>
    public static string ControllerUrl(string rawAddress)
    {
        var address = rawAddress.Trim();
        if (address.StartsWith(":", StringComparison.Ordinal))
        {
            address = "127.0.0.1" + address;
        }
        else if (address.StartsWith("*:", StringComparison.Ordinal))
        {
            address = "127.0.0.1:" + address[2..];
        }
        else if (address.StartsWith("0.0.0.0:", StringComparison.Ordinal))
        {
            address = "127.0.0.1:" + address["0.0.0.0:".Length..];
        }
        else if (address.StartsWith("[::]:", StringComparison.Ordinal))
        {
            address = "127.0.0.1:" + address["[::]:".Length..];
        }

        if (!address.Contains("://", StringComparison.Ordinal))
        {
            address = "http://" + address;
        }

        if (!Uri.TryCreate(address, UriKind.Absolute, out var uri) ||
            string.IsNullOrEmpty(uri.Host) ||
            uri.IsDefaultPort)
        {
            throw new ClashIntegrationException(
                new ClashIntegrationError.InvalidControllerAddress(rawAddress));
        }

        var host = uri.Host.ToLowerInvariant();
        if (host is not ("127.0.0.1" or "localhost" or "::1"))
        {
            throw new ClashIntegrationException(new ClashIntegrationError.UnsafeControllerHost(host));
        }

        // SchemeAndServer 恒不携带路径 → authority-only 地址无尾斜杠（与 Swift absoluteString 一致）；
        // 显式携带路径前缀的控制器地址保留路径（GetComponents 的 Path 分量为转义后的原始形态）。
        var server = uri.GetComponents(UriComponents.SchemeAndServer, UriFormat.UriEscaped);
        var path = uri.AbsolutePath;
        return path is "/" or "" ? server : server + path;
    }

    private sealed record CandidateConfiguration(string Path, string ClientName);
}
